bsip-0047.md

```diff
@@ -0,0 +1,182 @@
+* If any extension is present, `account_options.voting_account == GRAPHENE_PROXY_PER_CATEGORY_ACCOUNT`
+
+
+### `account_object`(Existing)
+
+**Note:** this is an implementation detail and meant to be a hint for later development. The actual implementation may differ.
+
+The account object shall receive these new fields:
+
+* `flat_set<vote_id_type> committee_votes`
+* `flat_set<vote_id_type> witness_votes`
+* `flat_set<vote_id_type> worker_votes`
+
+`account_create_operation` and `account_update_operation` shall be modified to create these sets by filtering the `account_options.votes` field appropriately. `account_update_votes_operation` shall also create these sets by filtering the `votes_to_add` and `votes_to_remove` fields appropriately.
+
+
+### `account_update_votes_operation` (New)
+
+The following is an example implementation.
+
+```
+struct account_update_votes_operation : public base_operation
+{
+	asset fee;
+
+	/// The account to update
+	account_id_type account;
+
+	/// New account options
+	flat_set<vote_id_type>    votes_to_add;
+	flat_set<vote_id_type>    votes_to_remove;
+
+	// A new voting account to set
+	optional<account_id_type> committee_voting_account
+	optional<account_id_type> witness_voting_account
+	optional<account_id_type> worker_voting_account
+
+	// A new number of witness_votes to set
+	optional<uint16_t> num_witness;
+
+	// A new number of committee_votes to set
+	optional<uint16_t> num_committee;
+
+	// For future extensions (see account_update_operation)
+	extensions_type extensions;
+};
+```
+
+Changes to direct votes can be specified by means of a *list* of vote identifiers to be added or removed. This has multiple advantages:
+
+* Reduces the size of the operation compared to replacing the entire slate
+* Allows to easily compare votes before and after the operation
+
+**Validation checks:**
+
+* `account` must be a valid account ID, must exist, must authorize the operation, must pay fee
+* `committee_voting_account`, if present, must be a valid account ID, must exist
+* `witness_voting_account`, if present, must be a valid account ID, must exist
+* `worker_voting_account`, if present, must be a valid account ID, must exist
+* Validate the existence of each vote to add and remove
+* `num_witness`, if present, must be greater than or equal to  &ge; 0
+* `num_committee`, if present, must be greater than or equal to  &ge; 0
+
+**Evaluation:**
+
+* Update the account's `account_options.committee_votes`, `account_options.witness_votes`, and `account_options.worker_votes` by filtering the `votes_to_add` and `votes_to_remove` field appropriately
+* Update the `account_options.num_witness` if `witness_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT` and if `num_witness` is present
+* Update the `account_options.num_committee` if `committee_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT` and if `num_committee` is present
+* Update the account's last vote time to the head block time
+
```

Comment:
* If any of `committee_voting_account`, `witness_voting_account` or `worker_voting_account` is present
    * update the corresponding field in the account's `account_options`
    * set the account's `account_options.voting_account` to `GRAPHENE_PROXY_PER_CATEGORY_ACCOUNT`